examples/barchart.rs

```diff
@@ -13,290 +13,320 @@
-    style::{Color, Modifier, Style},
-    terminal::{Frame, Terminal},
-    text::{Line, Span},
-    widgets::{Bar, BarChart, BarGroup, Block, Paragraph},
+    crossterm::event::{self, Event, KeyCode},
+    layout::{Constraint, Direction, Layout},
+    style::{Color, Style},
+    text::Line,
+    widgets::{Bar, BarChart, BarGroup, Block, Borders},
+    Frame,
 };
 
-struct Company<'a> {
-    revenue: [u64; 4],
-    label: &'a str,
-    bar_style: Style,
+use self::terminal::Terminal;
+
+const COMPANY_COUNT: usize = 3;
+const PERIOD_COUNT: usize = 4;
+
+struct App {
+    should_exit: bool,
+    data: Vec<Bar<'static>>,
+    companies: [Company; COMPANY_COUNT],
+    revenues: [Revenues; PERIOD_COUNT],
 }
 
-struct App<'a> {
-    data: Vec<(&'a str, u64)>,
-    months: [&'a str; 4],
-    companies: [Company<'a>; 3],
+struct Revenues {
+    period: &'static str,
+    revenues: [u32; COMPANY_COUNT],
 }
 
-const TOTAL_REVENUE: &str = "Total Revenue";
+struct Company {
+    short_name: &'static str,
+    name: &'static str,
+    color: Color,
+}
 
-impl<'a> App<'a> {
+fn main() -> Result<()> {
+    let mut terminal = terminal::init()?;
+    let app = App::new();
+    app.run(&mut terminal)?;
+    terminal::restore_terminal()?;
+    Ok(())
+}
+
+impl App {
     fn new() -> Self {
-        App {
-            data: vec![
-                ("B1", 9),
-                ("B2", 12),
-                ("B3", 5),
-                ("B4", 8),
-                ("B5", 2),
-                ("B6", 4),
-                ("B7", 5),
-                ("B8", 9),
-                ("B9", 14),
-                ("B10", 15),
-                ("B11", 1),
-                ("B12", 0),
-                ("B13", 4),
-                ("B14", 6),
-                ("B15", 4),
-                ("B16", 6),
-                ("B17", 4),
-                ("B18", 7),
-                ("B19", 13),
-                ("B20", 8),
-                ("B21", 11),
-                ("B22", 9),
-                ("B23", 3),
-                ("B24", 5),
-            ],
-            companies: [
-                Company {
-                    label: "Comp.A",
-                    revenue: [9500, 12500, 5300, 8500],
-                    bar_style: Style::default().fg(Color::Green),
-                },
-                Company {
-                    label: "Comp.B",
-                    revenue: [1500, 2500, 3000, 500],
-                    bar_style: Style::default().fg(Color::Yellow),
-                },
-                Company {
-                    label: "Comp.C",
-                    revenue: [10500, 10600, 9000, 4200],
-                    bar_style: Style::default().fg(Color::White),
-                },
-            ],
-            months: ["Mars", "Apr", "May", "Jun"],
+        Self {
+            should_exit: false,
+            data: Self::random_barchart_data(),
+            companies: Company::fake_companies(),
+            revenues: Revenues::fake_revenues(),
```

Comment:
(nonblocking)

I have a minor preference for keeping all the data together. What do you think about `Self::fake_companies()` and `Self::fake_revenues()` instead and having the implementations for these functions live next to the implementation for `Self::random_barchart_data()`?

Reply:
I thnk you're right about moving these out of companies and revenues, but I land on them being bare functions instead. 765c50929